SugarRecord/Source/Reactive/ReactiveCocoa/ReactiveStorage+ReactiveCocoa.swift

```diff
@@ -6,40 +6,48 @@ public extension Storage {
     
     // MARK: - Operation
     
-    func rac_operation(op: (context: Context, save: () -> Void) throws -> Void) -> SignalProducer<Void, Error> {
+    func rac_operation<T>(op: (context: Context, save: () -> Void) throws -> T) -> SignalProducer<T, Error> {
         return SignalProducer { (observer, disposable) in
             do {
-                try self.operation { (context, saver) throws in
+                let returnedObject = try self.operation { (context, saver) throws in
                     try op(context: context, save: {
                         saver()
                     })
-                    observer.sendCompleted()
                 }
+                
+                observer.sendNext(returnedObject)
+                observer.sendCompleted()
+
             }
             catch {
                 observer.sendFailed(Error.Store(error))
             }
         }
     }
     
-    func rac_operation(op: (context: Context) throws -> Void) -> SignalProducer<Void, Error> {
+    func rac_operation<T>(op: (context: Context) throws -> T) -> SignalProducer<T, Error> {
         return self.rac_operation { (context, saver) throws in
-            try op(context: context)
+            let returnedObject = try op(context: context)
             saver()
+            
+            return returnedObject
         }
     }
     
-    func rac_backgroundOperation(op: (context: Context, save: () -> Void) throws -> Void) -> SignalProducer<Void, Error> {
+    func rac_backgroundOperation<T>(op: (context: Context, save: () -> Void) throws -> T) -> SignalProducer<T, Error> {
         return SignalProducer { (observer, disposable) in
             let priority = DISPATCH_QUEUE_PRIORITY_DEFAULT
             dispatch_async(dispatch_get_global_queue(priority, 0)) {
                 do {
-                    try self.operation { (context, saver) throws in
+                    let returnedObject = try self.operation { (context, saver) throws in
                         try op(context: context, save: {
                             saver()
                         })
-                        observer.sendCompleted()
                     }
+
+                    observer.sendNext(returnedObject)
```

Comment:
I think it is accepted practice for the subscriber to observe on the thread
of there choice.

On Sun, Jun 12, 2016 at 23:55 Pedro Piñera Buendía <notifications@github.com>
wrote:

> In
> SugarRecord/Source/Reactive/ReactiveCocoa/ReactiveStorage+ReactiveCocoa.swift
> <https://github.com/pepibumur/SugarRecord/pull/257#discussion_r66746118>:
>
> >                      }
> > +
> > +                    observer.sendNext(returnedObject)
>
> Should we return the object int he main thread?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pepibumur/SugarRecord/pull/257/files/3fba7a3afdeaf6783b334b177d068a22e017a448#r66746118>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAk6dJ1Sgmkhk985qpSpjcr6s9NLnzYjks5qLP7FgaJpZM4I0AcQ>
> .
>